xray/commands/curation/analyse.go

```diff
@@ -0,0 +1,170 @@
+package curation
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/jfrog/jfrog-cli-core/v2/utils/coreutils"
+	utils2 "github.com/jfrog/jfrog-cli-core/v2/xray/commands/utils"
+	"github.com/jfrog/jfrog-client-go/artifactory"
+	"github.com/jfrog/jfrog-client-go/auth"
+	"github.com/jfrog/jfrog-client-go/utils/io/httputils"
+	"github.com/jfrog/jfrog-client-go/utils/log"
+	"github.com/jfrog/jfrog-client-go/xray/services/utils"
+	"net/http"
+	"strings"
+)
+
+type treeAnalyzer struct {
+	artiManager       artifactory.ArtifactoryServicesManager
+	artAuth           auth.ServiceDetails
+	httpClientDetails httputils.HttpClientDetails
+	url               string
+	repo              string
+	tech              coreutils.Technology
+}
+
+const (
+	BlockMessageKey  = "jfrog packages curation"
+	NotBeingFoundKey = "not being found"
+)
+
+func (nc *treeAnalyzer) recursiveNodeCuration(graph *utils.GraphNode, respStatus *[]PackageStatus, parent, parentVersion string, isRoot bool) error {
+	if parent == "" && !isRoot {
+		_, parent, parentVersion = getUrlNameAndVersionByTech(nc.tech, graph.Id, nc.url, nc.repo)
+	}
+	for _, node := range graph.Nodes {
+		packageUrl, name, version := getUrlNameAndVersionByTech(nc.tech, node.Id, nc.url, nc.repo)
+		resp, _, err := nc.artiManager.Client().SendHead(packageUrl, &nc.httpClientDetails)
+		if err != nil || resp.StatusCode >= 400 {
+			if resp == nil {
+				return err
+			}
+			if resp.StatusCode != http.StatusForbidden {
+				log.Error(fmt.Sprintf("Failed to head package %s for download url: %s, status-code: %v, err: %v", node.Id, packageUrl, resp.StatusCode, err))
+				continue
+			}
+		}
+		if resp.StatusCode == http.StatusForbidden {
+			err = nc.addBlockedPackage(packageUrl, parent, parentVersion, respStatus, name, version)
+			if err != nil {
+				log.Error(fmt.Sprintf("%v", err))
+				continue
+			}
+		}
+
+		if err := nc.recursiveNodeCuration(node, respStatus, parent, parentVersion, false); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (nc *treeAnalyzer) addBlockedPackage(packageUrl string, parent, parentVersion string, respStatus *[]PackageStatus, name string, version string) error {
+	getResp, respBody, _, err := nc.artiManager.Client().SendGet(packageUrl, true, &nc.httpClientDetails)
+	if err != nil {
+		if getResp == nil {
+			return err
+		}
+		if getResp.StatusCode != http.StatusForbidden {
+			log.Error(fmt.Sprintf("Failed to get package %s, version: %s for download url: %s. err: %v", name, version, packageUrl, err))
```

Comment:
Since this message is very close to another message in the `recursiveNodeCuration` function, let's use a const with a format for both.